### PR TITLE
Fix HTTP error handling in remote.Client.Store()

### PR DESCRIPTION
**Pull request by juliusv**
Regression introduced in
https://github.com/prometheus/prometheus/commit/e5d7bbfc3c7e38b4ee257dd1c6bc9b3e28b23d57

### CODE REVIEW

**Comment by juliusv** [comment]
@fabxc @tomwilkie

File: storage/remote/client_test.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+	for i, test := range tests {
+		server := httptest.NewServer(
+			http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				http.Error(w, "test error", test.code)
+			}),
+		)
+
+		serverURL, err := url.Parse(server.URL)
+		if err != nil {
+			panic(err)
```
Comment (tomwilkie) [inline storage/remote/client_test.go]: Could use `t.Fatalf` instead?

File: storage/remote/client_test.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+		}
+
+		c, err := NewClient(0, &clientConfig{
+			url:     &config.URL{serverURL},
+			timeout: model.Duration(time.Second),
+		})
+
+		err = c.Store(nil)
+		if test.shouldFail {
```
Comment (tomwilkie) [inline storage/remote/client_test.go]: Should also test `err == nil` if !test.shouldFail ?

File: storage/remote/client_test.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/config"
+)
+
+func TestStoreHTTPErrorHandling(t *testing.T) {
+	tests := []struct {
+		code        int
+		shouldFail  bool
+		recoverable bool
```
Comment (tomwilkie) [inline storage/remote/client_test.go]: You could pass in the expected error as part of the test case, then the checks would be a single `reflect.DeepEqual`.

**Comment by tomwilkie** [comment]
👍 from me, couple of minor nits on the test.

Travis failure is due to license check... which is weird.

**Comment by brian-brazil** [comment]
You need to add the legal stuff to the new test file.

File: storage/remote/client_test.go (hunk 1 above)
Comment (juliusv) [inline storage/remote/client_test.go]: I'd keep this a panic, as it should never fail, as it would be a bug in the test code.

**Comment by juliusv** [comment]
Fixed all the things. Just doing a `DeepReflect` on `err` makes the code much simpler, although it now requires an exact replication of the error strings...

**Comment by tomwilkie** [comment]
> Just doing a DeepReflect on err makes the code much simpler, although it now requires an exact replication of the error strings...

You could extract the error generation out into a function if you liked.

👍

**Comment by juliusv** [comment]
> You could extract the error generation out into a function if you liked.

Meh, overkill IMO...

**Pull request merged**